operator/builtin/parser/csv/csv.go

```diff
@@ -80,47 +101,77 @@ func (c CSVParserConfig) Build(context operator.BuildContext) ([]operator.Operat
 type CSVParser struct {
 	helper.ParserOperator
 	header          string
+	headerLabel     string
+	dynamicHeader   bool
 	headerDelimiter rune
 	fieldDelimiter  rune
 	numFields       int
+	parse           ParseFunc
 }
 
 // Process will parse an entry for csv.
-func (r *CSVParser) Process(ctx context.Context, entry *entry.Entry) error {
-	return r.ParserOperator.ProcessWith(ctx, entry, r.parse)
-}
-
-// parse will parse a value using the supplied csv header.
-func (r *CSVParser) parse(value interface{}) (interface{}, error) {
-	var csvLine string
-	switch t := value.(type) {
-	case string:
-		csvLine += t
-	case []byte:
-		csvLine += string(t)
-	default:
-		return nil, fmt.Errorf("type '%T' cannot be parsed as csv", value)
+func (r *CSVParser) Process(ctx context.Context, e *entry.Entry) error {
+	if r.dynamicHeader {
+		h, ok := e.Labels[r.headerLabel]
+		if !ok {
+			// TODO: returned error is not logged, so log it here
+			err := fmt.Errorf("failed to read dynamic header label %s", r.headerLabel)
+			r.Error(err)
+			return err
+		}
+		r.parse = generateParseFunc(h, r.headerDelimiter, r.fieldDelimiter, r.numFields)
 	}
+	return r.ParserOperator.ProcessWith(ctx, e, r.parse)
+}
 
-	reader := csvparser.NewReader(strings.NewReader(csvLine))
-	reader.Comma = r.fieldDelimiter
-	reader.FieldsPerRecord = r.numFields
-	parsedValues := make(map[string]interface{})
-
-	for {
-		record, err := reader.Read()
-		if err == io.EOF {
-			break
+type ParseFunc func(interface{}) (interface{}, error)
+
+// generateParseFunc returns a parse function for a given header, allowing
+// each entry to have a potentially unique set of fields when using dynamic
+// field names retrieved from an entry's label
+func generateParseFunc(header string, headerDelimiter, fieldDelimiter rune, numFields int) ParseFunc {
+	return func(value interface{}) (interface{}, error) {
+		var csvLine string
+		switch t := value.(type) {
+		case string:
+			csvLine += t
+		case []byte:
+			csvLine += string(t)
+		default:
+			return nil, fmt.Errorf("type '%T' cannot be parsed as csv", value)
 		}
 
-		if err != nil {
-			return nil, err
+		reader := csvparser.NewReader(strings.NewReader(csvLine))
+		reader.Comma = fieldDelimiter
+		reader.FieldsPerRecord = numFields
+		parsedValues := make(map[string]interface{})
+
+		for {
+			record, err := reader.Read()
+			if err == io.EOF {
+				break
+			}
+
+			if err != nil {
+				return nil, err
+			}
+
+			headerFields := strings.Split(header, string([]rune{headerDelimiter}))
+
+			// When numFields is less than 1 (due to dynamic header detection) we need to check if
+			// the correct number of fields exists before looping. When numFields is greater than 0,
+			// encoding/csv will return its own error when this mismatch occurs.
+			if numFields < 1 {
+				if len(headerFields) != len(record) {
+					return nil, fmt.Errorf("number of fields does not match number of fields defined by the header '%s'", headerFields)
+				}
+			}
```

Comment:
Thinking through what we're optimizing here, and what we're not:
- We're precalculating `numFields`, so that we don't have to recalculated it for every entry in the static case.
- We're splitting the header labels every time no matter what, which is a much more costly operation.

It's a bit tricky to optimize for both scenarios at the same time, so I suggest we consider generating parse funcs separately for the two cases: 
```go
func staticParseFunc(header string, headerDelimiter, fieldDelimiter rune) ParseFunc {
	return func(value interface{}) (interface{}, error) {
		var csvLine string
		switch t := value.(type) {
		case string:
			csvLine += t
		case []byte:
			csvLine += string(t)
		default:
			return nil, fmt.Errorf("type '%T' cannot be parsed as csv", value)
		}

		headerFields := strings.Split(header, string([]rune{headerDelimiter}))

		reader := csvparser.NewReader(strings.NewReader(csvLine))
		reader.Comma = fieldDelimiter
		reader.FieldsPerRecord = len(headerFields) // TODO remove `numFields` from operator struct
		parsedValues := make(map[string]interface{})

		for {
			record, err := reader.Read()
			if err == io.EOF {
				break
			}

			if err != nil {
				return nil, err
			}

			for i, key := range headerFields {
				parsedValues[key] = record[i]
			}
		}

		return parsedValues, nil
	}
}

func dynamicParseFunc(header string, headerDelimiter, fieldDelimiter rune) ParseFunc {
	return func(value interface{}) (interface{}, error) {
		var csvLine string
		switch t := value.(type) {
		case string:
			csvLine += t
		case []byte:
			csvLine += string(t)
		default:
			return nil, fmt.Errorf("type '%T' cannot be parsed as csv", value)
		}

		reader := csvparser.NewReader(strings.NewReader(csvLine))
		reader.Comma = fieldDelimiter
		// don't set reader.FieldsPerRecord
		parsedValues := make(map[string]interface{})

		for {
			record, err := reader.Read()
			if err == io.EOF {
				break
			}

			if err != nil {
				return nil, err
			}

			headerFields := strings.Split(header, string([]rune{headerDelimiter}))

			if len(headerFields) != len(record) {
				return nil, fmt.Errorf("number of fields does not match number of fields defined by the header '%s'", headerFields)
			}
			
			for i, key := range headerFields {
				parsedValues[key] = record[i]
			}
		}

		return parsedValues, nil
	}
}

```

This leaves some duplicated code, which isn't ideal, but the common parts can likely be extracted. I think it'd be ok to leave the duplication for now. What's most important is performance, and I think we'll see an improvement in the static case by taking this approach.

Reply:
I think it makes sense to remove numFields, and keep a single generate function. The way I have it (recent commit) does the following
- get number of fields by splitting the header
- set csv's FieldsPerRecord to headerFields.
  - In the case of dynamic headers, this happens on every entry.  If we do not set it, csv's package will detect it anyway.
- removed the check for number of fields not matching number of records / columns in the entry. CSV package does this for us and gives an appropriate error.

Reply:
Definitely an improvement. 

I'm not a huge fan of splitting the headers for every record, since in the static case this is unnecessary effort. Not a big enough deal to block this though.